VariantShredding.md

```diff
@@ -25,290 +25,316 @@
 
-If a value is an object, and the `object` field is present but does not contain all of the fields in the value, then any remaining fields are stored in an object in `variant_value`.
-In the Parquet example above, if field `b` was an object of the form `{"c": 1, "d": 2}"`, then the object `{"d": 2}` would be stored in `variant_value`, and the `c` field would be shredded recursively under `object.c`.
+#### Arrays
 
-Note that an array is always fully shredded if there is an `array` field, so the above consideration for `object` is not relevant for arrays: only one of `array` or `variant_value` may be non-null at a given level.
+Arrays can be shredded using a 3-level Parquet list for `typed_value`.
 
-# Using variant_value vs. typed_value
+If the value is not an array, `typed_value` must be null.
+If the value is an array, `value` must be null.
 
-In general, it is desirable to store values in the `typed_value` field rather than the `variant_value` whenever possible.
-This will typically improve encoding efficiency, and allow the use of Parquet statistics to filter at the row group or page level.
-In the best case, the `variant_value` fields are all null and the engine does not need to read them (or it can omit them from the schema on write entirely).
-There are two main motivations for including the `variant_value` column:
+The list `element` must be a required group that contains `value` and `typed_value` fields.
+The element's `value` field stores the element as Variant-encoded `binary` when the `typed_value` is not present or cannot represent it.
+The `typed_value` field may be omitted when not shredding elements as a specific type.
+When `typed_value` is omitted, `value` must be `required`.
 
-1) In a case where there are rare type mismatches (for example, a numeric field with rare strings like “n/a”), we allow the field to be shredded, which could still be a significant performance benefit compared to fetching and decoding the full value/metadata binary.
-2) Since there is a single schema per file, there would be no easy way to recover from a type mismatch encountered late in a file write. Parquet files can be large, and buffering all file data before starting to write could be expensive. Including a variant column for every field guarantees we can adhere to the requested shredding schema.
+For example, a `tags` Variant may be shredded as a list of strings using the following definition:
+```
+optional group tags (VARIANT) {
+  required binary metadata;
+  optional binary value;
+  optional group typed_value (LIST) {   # must be optional to allow a null list
+    repeated group list {
+      required group element {          # shredded element
+        optional binary value;
+        optional binary typed_value (STRING);
+      }
+    }
+  }
+}
+```
 
-# Top-level metadata
+All elements of an array must be present (not missing) because the `array` Variant encoding does not allow missing elements.
```

Comment:
I might have read to quickly why is this true?  it seems like ["foo", null, "bar"] should be encodable

Reply:
I think the answer here is me reading imprecisely.  I think we have a lot of technical terms and precision matters, a glossary to distinguish them I think would help readers better understand the spec:

So far I think:
1.  Field
2. Column
3. Null value (explicit null in the variant)
4. Missing (implicit null in variant object).
5. null/undefined values according to parquet definition levels (for shredding these are always used to indicate missing values or a a desciminator for which how to interpreted typed_value and values fields).

Reply:
"optional", "required" vs present in the schema is another area that I think might have caused confusion.